core/src/main/scala/org/apache/spark/deploy/rest/mesos/MesosRestServer.scala

```diff
@@ -99,7 +99,11 @@ private[mesos] class MesosSubmitRequestServlet(
     // cause spark-submit script to look for files in SPARK_HOME instead.
     // We only need the ability to specify where to find spark-submit script
     // which user can user spark.executor.home or spark.home configurations.
+    //
+    // Due to https://issues.scala-lang.org/browse/SI-6654, `filterKeys` returns an
+    // unserializable object, so we must call `.map(identity)` on the result.
     val environmentVariables = request.environmentVariables.filterKeys(!_.equals("SPARK_HOME"))
+      .map(identity)
```

Comment:
Actually, there's a slightly nicer way...
```
// Do not use `filterKeys` here to avoid SI-6654, which breaks ZK persistence
val environmentVariables = request.environmentVariables
  .filter { case (k, _) => k != "SPARK_HOME" }
```